### ISSUE TITLE
No error in authState when error from redirect.

### ISSUE THREAD

**Issue by sdontas** [okta/okta-auth-js#550]
Hello! My team and I are using `@okta/okta-auth-js@4.2.0` along with `@okta/okta-react@4.0.0` in order to create an application with authentication. `@okta/okta-react` got released 3 days ago and we wanted to upgrade in order to be up to date with the latest releases.

While migrating to the new releases, we found out the following:

- When a user that is not assigned to the application tries to login, then the `error` property in `authState` is not set at all. The url contains the errors as query parameters. For reference the query parameters look like this:
`?state=HjbBu...ga1F&error=access_denied&error_description=User+is+not+assigned+to+the+client+application.`
However, the `authState` emitted does not contain any `error` property.

For reference, our use case is the simplest one found in [`@okta/okta-react`](https://github.com/okta/okta-react/blob/okta-react-4.0.0/README.md#creating-react-router-routes-with-function-based-components).

**Technical**

`@okta/okta-react` uses the function [`handleLoginRedirect`](https://github.com/okta/okta-auth-js/blob/master/lib/browser/browser.ts#L490) in order to parse the tokens from the URL and redirect to where the user wanted to go. However, this function calls [`isLoginRedirect`](https://github.com/okta/okta-auth-js/blob/master/lib/browser/browser.ts#L518) before parsing the tokens from the URL, which in turn returns false if no `code` is found in the URL query parameters. In our case, because there are errors and no `code`, the function [`storeTokensFromRedirect`](https://github.com/okta/okta-auth-js/blob/master/lib/browser/browser.ts#L462) in the next line is not executed.

Further digging through the source code of both packages, we came to the conclusion that the `error` property is only set if the [`transformAuthState` configuration function throws an error](https://github.com/okta/okta-auth-js/blob/7d9bde387a38fd01f1d82d0d6f2385cbdf4ac6b9/lib/AuthStateManager.ts#L171).

**Note**

In general, we would expect that the `error` property would be populated when something like the above case happens (at least this is what happened in the previous major versions). We are wondering whether we are doing something wrong on our side, or if this is actually a bug in any of the repositories.

Thank you for your time in advance.

**Comment by shuowu** [comment]
@sdontas Thanks for reporting the issue! 

The `AuthStateManager` (in auth-js) currently only handles errors thrown from custom callbacks, like `transformAuthState`. For your use case, a workaround can be to provide your own [`LoginCallback`](https://github.com/okta/okta-react/blob/master/src/LoginCallback.tsx), then catch and handle the error from `oktaAuth.handleLoginRedirect`.

Internal Ref: OKTA-348505

**Comment by sdontas** [comment]
@shuowu Thank you for the reply.

>For your use case, a workaround can be to provide your own [`LoginCallback`](https://github.com/okta/okta-react/blob/master/src/LoginCallback.tsx), then catch and handle the error from `oktaAuth.handleLoginRedirect`.

Yes, this is the workaround we have come up with and are successfully using at the moment.

>The AuthStateManager (in auth-js) currently only handles errors thrown from custom callbacks, like transformAuthState.

Should we expect this behavior to change in future releases?

**Comment by shuowu** [comment]
@sdontas 
> Should we expect this behavior to change in future releases?

The error state would be handled in the future release, and as it just properly adds the `error` state, the change would just be a minor release (**non-breaking** change)

**Comment by sdontas** [comment]
@shuowu Thank you very much for your quick replies! Looking forward to it.

**Comment by minkeshj** [comment]
Facing the same issue. Any update on this ?

**Comment by phyolim** [comment]
> @sdontas Thanks for reporting the issue!
> 
> The `AuthStateManager` (in auth-js) currently only handles errors thrown from custom callbacks, like `transformAuthState`. For your use case, a workaround can be to provide your own [`LoginCallback`](https://github.com/okta/okta-react/blob/master/src/LoginCallback.tsx), then catch and handle the error from `oktaAuth.handleLoginRedirect`.
> 
> Internal Ref: [OKTA-348505](https://oktainc.atlassian.net/browse/OKTA-348505)

I am having the same issue. Could you clarify what you mean by handle the error from `oktaAuth.handleLoginRedirect`?
I tried to catch the error from there like below, but, it always resolves even though the login redirect has an error message-

```oktaAuth
oktaAuth.handleLoginRedirect()
.then((value) => {
console.log(value);
})
.catch((err) => {
console.log(err);
});```

**Comment by shuowu** [comment]
@phyolim As @sdontas described in the report, the cause is 
> However, this function calls isLoginRedirect before parsing the tokens from the URL, which in turn returns false if no code is found in the URL query parameters. In our case, because there are errors and no code, the function storeTokensFromRedirect in the next line is not executed.

We are working on a fix for this issue now, before the new release, you can create a wrapper function of `oktaAuth.handleLoginRedirect` to check the error state from the redirectUrl. The code can be something like:
```
const handleLoginRedirect = async () => {
  // ... parse query params from the redirect URL
  if (params['error'] || params['error_description']) {
    await this.storeTokensFromRedirect();
  } else {
    await oktaAuth.handleLoginRedirect();
  }
}

**Comment by shuowu** [comment]
@phyolim PR #583 fixes the `isLoginRedirect ` issue.

**Comment by phyolim** [comment]
Thank you!

**Comment by shuowu** [comment]
Fixed in v4.6.0. Closing this issue, feel free to re-open it if the issue still exist.
